### DISCUSSION TITLE
Teleop with with joystick-like input

### DISCUSSION THREAD

**Discussion by jasonimercer** [foxglove/community#21] [Q&A]
Would it be possible to add an advanced teleop panel that allows the control of a robot with a draggable nub? 

Options would be:
* Square limits or Circle limits. This would define the limits you could drag the nub. Square limits allows all combinations of the up-down value and the left-right value. Circle would limit max up-down based on the value of left-right and vice-versa. Releasing the nub would reset to 0.
* Dead zone sizes. These would define ranges where values from the "other" coordinate would be clamped to zero. This can make a pure turn in place easier or a pure "go forward" easier. Dead zones would show up somehow on the UI and a dead zone of 0 would remove the hint from the UI.
* "Tank" vs "Car" mode. "Tank" mode would make each coordinate decoupled. "Car" mode would negare the left-right when the up-down goes negative.
* Max/Mins on each coordinate.
* Map coordinates to linear-xyz / angular-xyz similar to your traditional teleop settings.

With the above it becomes possible to teleop a robot with finesse. The current teleop is all gas or all turns and when you factor in concepts like "about to move" low-level constraints, you get nearly uncontrollable surprises.

**Comment by amacneil** [comment]
@rgov how is your gamepad extension going? can we add it to our marketplace?

**Comment by rgov** [comment]
It is in early alpha [here](https://github.com/ARMADAMarineRobotics/studio-extension-gamepad) and I welcome feedback on that issue tracker, but I think it's premature to release it on the marketplace just yet.

I would love to see #3413 or a comparable solution land to clear up some hacks in the code as I move forward with supporting more types of controllers (currently it bakes in the Sixaxis).

**Comment by jasonimercer** [comment]
Ideally there would be a solution that could be used with a mouse and not rely on extra hardware. It is possible to walk around a facility behind a robot holding a laptop in your left arm and controlling the robot with your right hand. If there's a physical gamepad that requires 2 hands, you can no longer see what the robot sees but you still need to carry the laptop. Even if the gamepad drops down to single handed with careful balancing, you still would need to let go of the controller to click elsewhere to see extra information.

Gamepads are good input devices in some situations but are overhead on others.

**Comment by amacneil** [comment]
I agree, thanks for sharing this use case. It might be possible to support both input types (mouse/keyboard/joystick) in the same panel, or they could be separate panels.

This feature would be a good opportunity for an extension, because improving teleop is not currently on our core roadmap. We did recently launch support for [deploying extensions privately within an organization](https://foxglove.dev/docs/studio/extensions/publish#sharing-extensions-with-your-foxglove-organization), in addition to publishing open source extensions for everyone to access.

FYI I have converted the issue to a discussion since it is quite open-ended.

**Comment by joshnewans** [comment]
Hi guys, I was going to raise this as a feature request issue, when I came across this discussion, and I thought I'd provide my 2c having given it some thought. I'm keen to help get it off the ground if there's anything I can do, however my time and web dev experience are both fairly limited.

It would be great to see gamepad/joystick support added, so that bluetooth controllers such as those shown below, paired with cheap tablets, can make for a pretty slick but low-cost interface. Of course, having gamepad support in general is useful even when working at a laptop/desktop/etc.

![Gamepad](http://ipega.hk/uploads/201906/5cf75bf342039.jpg)

Such a panel would publish a `sensor_msgs/Joy` message, and it is up to the software beyond that to interpret that message as it sees fit - which it may already be doing for existing teleop procedures. Rather than being restricted to a `Twist` message, these controls can be mapped to any desired functionality.

I imagine the development roadmap looking something like:

1. Initial support for connection to a generic gamepad and publishing a `Joy` message. Display the buttons/axes being pressed on the screen for feedback.
2. *"View" mode* - that can instead subscribe to a `Joy` topic and display the pressed buttons. This can be used e.g. for playing back recorded logs that were teleoperated, and be aware of what the operator was doing.
3. *Configurable button layout* - this could include setting size/shape/location of buttons, and axes. Also, the ability to represent axis pairs as a thumbstick (or as individual horizontal/vertical bars)
4. *Simulated gamepad mode* - Turn those feedback elements into clickable buttons and draggable thumbsticks, for mouse or touchscreen interface.
5. *Keyboard control* - Ability to map keypresses to those simulated buttons/sticks

Not a trivial task, but it seems to me to be a fairly logical approach that results in most of the functionality people would want, in a single, unified panel.

Unaware of the above works, I went about converting some old demo code I wrote a few years ago into a Foxglove extension and it worked great! [Repo here](https://github.com/joshnewans/joy-panel)
It currently uses a (possibly outdated) npm package for handling the gamepad input, but I believe it is a wrapper for the [official HTML Gamepad API](https://developer.mozilla.org/en-US/docs/Web/API/Gamepad_API/Using_the_Gamepad_API) which should work with any gamepad.

There is a video demo of its current state [here](https://youtu.be/0ZW7KBM8LkY).

For what it's worth, [here's](https://youtu.be/hkkG-Sgi9Sk?t=378) a recent tutorial I made, which included a quick demo of Foxglove.
Thanks all for your hard work!

**Reply by jtbandes** [reply]
Thanks for sharing, these tutorials are great! 🙌 

You'd be welcome to submit a PR adding your extension to the marketplace so it's more easily discoverable: https://github.com/foxglove/studio-extension-marketplace

Aside: one minor comment I had watching the foxglove section is that you recommended running rosapi_node manually. Usually the way folks launch rosbridge is using [this launch file](https://github.com/RobotWebTools/rosbridge_suite/blob/ros2/rosbridge_server/launch/rosbridge_websocket_launch.xml) (i.e. `ros2 launch rosbridge_server rosbridge_websocket_launch.xml`) which includes both the websocket server and the rosapi node. 

Also please stay tuned at https://github.com/foxglove/ros-foxglove-bridge where we are building a rosbridge_server replacement that will provide much improved performance (being written in C++ rather than Python like the current rosbridge) and richer integration with Foxglove Studio.

**Reply by joshnewans** [reply]
Thanks Jacob, all three of those are very helpful feedback points I wasn't aware of!

I've not had a chance to really use Foxglove "in anger" yet (either in my hobby work or my day job), but every time I see an update or blog post it seems to be something I've always thought "it would be great if there was a standard tool for X", so I'm keen to see it keep growing :D

**Comment by rgov** [comment]
Hi @joshnewans I wonder if you saw my extension at https://github.com/ARMADAMarineRobotics/studio-extension-gamepad ? Admittedly it is in alpha right now but it sounds very much like what you describe — I’m even using the same hardware :)

I haven’t worked on it in a while but it supports multiple game pad configurations, it sends  `Joy` messages, it simulates input when in playback mode, etc.

If you’d like to collaborate I’d love PRs to get it to beta and released on the marketplace.

**Reply by joshnewans** [reply]
Hey Ryan! Yeah that's what I was referring to when I said "unaware of the works above" - I hadn't seen your code until I found this thread this morning. I also saw your comment about it being tied to PlayStation controllers and wasn't sure how heavy that limitation was, I planned to try it out some time.

Sounds like you've got a lot of what I was thinking already in there so I'm keen to give it a go and contribute if I can.

For what it's worth, the four controllers I have that I swap between are:
- Xbox 360 wired
- Xbox One S Bluetooth
- Generic 2.4GHz USB PlayStation layout (shown in the video)
- iPEGA 9083S Tablet/phone BT controller (pictured in my post on this thread)

**Reply by rgov** [reply]
Sorry I’m viewing on mobile and didn’t see the whole thread.

I started with PS3 because it was natively supported on macOS. Then Apple switched from supporting PS3 to supporting Xbox. But if you’re using the Gamepad API, whether it’s a USB or Bluetooth HID device or a IAP accessory it should work (*in MobileSafari).
